### PR TITLE
[ShapeOpt] Adding folder structure creation

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This PR adds the posibility to create folder structure for each response and iterations. This is usefull in doing optimization in fluid problems. This does not change the existing behaviour.

**Changelog**
- Added interface to core `response_function_interface` to set optimization model part
- ~~Added folder structure settings to `model_part_controller_factory`~~
- Added folder structure creation to `analyzer_internal`

### CODE REVIEW

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: i think this check is not needed, since you set this variable in the constructor?

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: same

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I see now that distinguishing between objective and constraint is not easily possible here, so the idea of the counter does not work that well. I would use the `identifier` (key of self.response_functions) instead.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I suggest to always initialize the members, but with None if no folder name is given.
Then you don't need to do the "hasattr" check below but can check for None which is more pythonic i think.

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: `output` is not the best term here since it might be confused with settings of the optimization output. In the end we do all the evaluation in this folder. 

Maybe the terms like `working_directory` or `iteration_scope` or `working_scope` fit better. 
I did not come up if something better either

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [inline kratos/python_scripts/response_functions/response_function_interface.py]:
```suggestion
    def SetEvaluationModelPart(self, optimization_model_part):
```
or 
```suggestion
    def SetAnalysisModelPart(self, optimization_model_part):
```

**Review by armingeiser — COMMENTED** [review]
Thanks @sunethwarna, i will need to test this with my applications as well, but the way you made it such it can be switched on/off is nice!

**Comment by sunethwarna** [comment]
Thanks for the feedback @armingeiser . I have another small question. When this is used with constraints where folder structure is not required, it still creates the folder structure with empty folders. I would like to have another method in the `response_function_interface` stating "IsFolderStructureRequired" so by default it is always false, if a constraint or reponse requires it, then it will write the corresponding folder structure.

What do you think?

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: not exactly. I only set the scope if folder structure is required only.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: not exactly. I only set the scope if folder structure is required only.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: i'll change :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I made it like that to have minimal cost in the case if this folder structure is not required.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: oh i wanted to delete this comment. please ignore

**Comment by armingeiser** [comment]
I think that makes sense. Then we can decide based on the response if we scope in or not. BEcause the response anyway needs to know what to do with the scope.

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (sunethwarna) [reply kratos/python_scripts/response_functions/response_function_interface.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: removed `hasattr`

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: removed `hasattr`

**Comment by sunethwarna** [comment]
I did the changes. Could you have a look @armingeiser ?

**Comment by armingeiser** [comment]
It looks god for me. I suggest to wait a bit with merging to see if someone else has comments.

**Review by armingeiser — DISMISSED** [review]

**Comment by dbaumgaertner** [comment]
Lemme also have a look at it. I will do so tomorrow or the day after.

**Comment by sunethwarna** [comment]
@dbaumgaertner that is totally fine :) Thanks in advance :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: Without further insight, that comment is inconsistent with the name of the following function (optimization vs evaluation part) and does not add any information.  Just remove it here.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: Can you explain to me, why and in which situations this function is needed?

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: I don't quite understand this directory. If I understood it correctly, then the iteration directory is a directory under the output_directory, which contains subdirs for each iteration. If yes, then is there really a need to specify the name for this top dir, or could it be just "optimization_iterations" by default - especially considering that one also can specify the output_dir?  One could also think of renaming the default into "design_iterations", something a bit more descriptive and different compared to the output_directory. But your call.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:34]: I actually like the idea of having a separate folder for each design. Question @sunethwarna @armingeiser : How do you plan to treat the external analyzer then? More precisely, having this iteration scope means that an external analyzer would also need to implement the IterationScope things? So, one would need to copy every time from the internal analyzer to an external analyzer, which not makes the definition of an external solver much more difficult, but also yields to code doubling. Have you thought about his?

**Review by dbaumgaertner — COMMENTED** [review]
Looks good in general and I also like the idea of the folder structure. Thanks a lot, @sunethwarna for making this effort, nice work! No major blocker from my side. But before approving, I would have some questions, which I posted at the respective lines. Another two questions:

1) does this change the default behaviour of the optimizer (are now always folders created)?
2) in line with the first question, do the changes affect the tests? From a quick test runs, it seems not?

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I'll remove

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: In the case of fluid optimizations, i need the updated mesh for the fluid solve. I cannot use the mesh from the optimization since I need to have time steps as well (RANS and URANS both have time steps (and of course fluid problems have different types of `elements` and `conditions` such as `FLUID_ELEMENTS`, `RANS_ELEMENTS`, `WALL_CONDITIONS`, `PERIODIC_CONDITIONS`), RANS ones are psedo time steps). So with this method it gives the opportunity to response function (in my case) to write the mdpa so, fluid simulations can read it.

I have implemented this in the following
https://github.com/KratosMultiphysics/Kratos/blob/0766630b840fb1f989e40ffc87ccdd6e5347d942/applications/RANSApplication/python_scripts/response_functions/lift_to_drag_response_function.py#L82-L95

In the case of fluid, cost of writing and reading mdpas is negligible compared to solve time of fluid primal problem.

Another reason why I write the mdpa (and final default assigned json file and other required files) is, I made the PR #8399 in such a way that, if someone wants to do further investigations with a specific iteration then they can copy the iteration and run that same iteration as a standalone simulation. This is also really important since sometimes I have to do post-processing (based on Kratos) also.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:34]: @dbaumgaertner As far as I think, external analyzers does not need to implement this. This is an internal scoping which is used only if a response requires it. If they don;t need it they don't need to implement it. (If the developers of the external analyzers need it, then they can take the `IterationScope` from the internal analyzer so we have less copy paste stuff)

As for the answers to main comments
> does this change the default behaviour of the optimizer (are now always folders created)?

No, it will create the folder structure if required by the response

https://github.com/KratosMultiphysics/Kratos/blob/0766630b840fb1f989e40ffc87ccdd6e5347d942/kratos/python_scripts/response_functions/response_function_interface.py#L59-L60

`IterationScope` will check whether reponse needs the folder strucutre. If a specific response function requires it, then folder structure will only be created for that response only. So we can have reponses with and without folder structure. (Tested it with constraints in the PR #8399 , and it works)

> in line with the first question, do the changes affect the tests? From a quick test runs, it seems not?

It does not change the tests nor change the performance of the tests. It only adds additionaly small overhead for checking `IsEvalautedInFolder` per each iteration (which is really a boolean check and negligigble)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I may extend: Until now we have assumed, that the internal responses share the optimization model part with the optimizer. This is not the case e.g. for the RANS response, since in every iteration we start an analysis (including import from an updated mdpa) from scratch. 

By passing the optimization model part, we have the chance to get the current design (shape update) from the optimizer. 

I am not completely happy with this solution yet, but we did not have a better idea until now.

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: I'm also in favor of removal of this setting and fixing it (We already have a fixed folder name "OptimizationResults", so this could be called "OptimizationIterations"). @armingeiser ?

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: :+1: from my side

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I just had another idea: 

What if the response would share the optimization_model_part with the optimizer (as it is done in the structural responses), but instead of directly using it for evaluation, it can create independent model_parts for primal/adjoint etc. 

Like that we would use the same logic everywhere and this function would not be needed. 

what do you think about:

```python
class RANSResponse():
    def __init__(...):
        self.design_model_part = model.GetSubModelPart(...) # this is expected to be imported by the optimizer
        self.primal_mp = ... # can be different from self.design_model_part, in that case the response needs to make sure to transfer updates of the design and gradients between the modelparts
        self.adjoint_mp = ... # if needed, also different from design_model_part
```

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: Or we rename the function to `def UpdateDesign(self, updated_design_mp, design_variables):`, that name is more descriptive to what is actually happening. 

The `design_variables` could be used to give a hint on what is updated (for ShapeOpt it will be SHAPE, but could also be properties), similar to how we expect gradients w.r.t a certain variable in the `GetGradient` function.

@MFusseder this might be relevant for you as well?

Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: I have a small question regarding the proposed `RANSResponse`. How can the response know which `ModelPart` is used in the `Model` for optimization. In this case either we have to pass the name of it via `json` or via function or we need to fix it. What do you think?

I like the idea of `UpdateDesign`, its more descriptive of what we doing, and we can pass design variables as well.

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: done :)

File: kratos/python_scripts/response_functions/response_function_interface.py
Comment (armingeiser) [inline kratos/python_scripts/response_functions/response_function_interface.py]:
```suggestion
    def UpdateDesign(self, updated_model_part, variable):
```

Comment (sunethwarna) [reply kratos/python_scripts/response_functions/response_function_interface.py]: Hmm... I am wondering how did the CI pass without this??

Comment (armingeiser) [reply kratos/python_scripts/response_functions/response_function_interface.py]: would it put all the calling arguments into a tuple for the single argument in the interface function? Not sure

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: Small change, but much better now!

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: Did you think about renaming it to "DesignIterations"? I would personally find it more descriptive and it would also a bit more distinguishable from the "OptimizationResults" folder. But as I said, your call after all.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:34]: Ah, understood. The fact that every response may decide on its own makes sense. This way, also an external analyzer can do whatever he likes.

**Review by dbaumgaertner — DISMISSED** [review]
Go from my side. But before merging, think about the name suggests I gave. This is a minor thing, though.

File: applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/model_part_controller_factory.py]: done :)

**Comment by sunethwarna** [comment]
All the suggestions are applied. Could maybe @armingeiser or @dbaumgaertner please re-approve?

**Review by armingeiser — APPROVED** [review]
Thanks Suneth!

**Pull request merged**
